### ISSUE TITLE
Setting different speeds for both directions can cause problems with loops

### ISSUE THREAD

**Issue by easbar** [graphhopper/graphhopper#1631]
Routing can fail if we use `MotorCycleEncoder` instead of `CarFlagEncoder` and there are loops (for OSM they are excluded during import though), because motorcycle allows setting different speeds for both directions. See this branch: [`motorcycle_edge_based_ch_bug`]( https://github.com/graphhopper/graphhopper/tree/motorcycle_edge_based_ch_bug)

**Comment by easbar** [comment]
The test only shows the problem for edge-based CH, but I am not sure if it could be a problem for other (edge-based ?) algorithms as well, not sure if this is fully covered by tests.

**Comment by easbar** [comment]
> The test only shows the problem for edge-based CH, but I am not sure if it could be a problem for other (edge-based ?) algorithms as well, not sure if this is fully covered by tests.

Yes, this is a problem for all edge-based algorithms and its nothing specific to edge-based CH. We cannot reliably use different speeds for the two directions for loops. The current solution is still that loops are excluded during the import. Here is a recent commit where `RandomizedRoutingTest` fails sometimes (run it many times): eda2b53e20c935e4d14cd5338f7baebe2321427d

Interestingly the test is 'fixed' by bbb15dfb6b67ace6e060c846ba69fae15641eddb, but this is only because with this commit the speeds aren't properly set to different values due to a changed `randomGraph` method, see this comment: https://github.com/graphhopper/graphhopper/pull/2202#discussion_r541996509

**Comment by easbar** [comment]
In our current model loops with different properties for the two directions are not well defined. Every edge has an edge ID, a base node ID (A) and an adj node ID (B). A loop is an edge where A=B. The direction of an edge is the relation between A and B, but since A=B a loop has only one direction. Therefore setting different edge properties for the 'fwd' and 'bwd' direction of a loop is just wrong. Currently there are a few places (like `DefaultEdgeFilter`) where we do special handling for loop edges, but these would not be necessary if we did not use direction-dependent properties for loop edges in the first place. Can we somehow deny setting different values for fwd/bwd directions on loops?

This should not be confused with loop shortcuts in edge-based CH. The direction of these loop shortcuts is actually well defined, because of the first/last orig edges.

**Comment by easbar** [comment]
Maybe we could replace, for example:

```java
       @Override
        public EdgeIteratorState set(BooleanEncodedValue property, boolean value) {
            property.setBool(reverse, getFlags(), value);
            baseGraph.writeFlags(edgePointer, getFlags());
            return this;
        }

        @Override
        public EdgeIteratorState setReverse(BooleanEncodedValue property, boolean value) {
            property.setBool(!reverse, getFlags(), value);
            baseGraph.writeFlags(edgePointer, getFlags());
            return this;
        }
}
```

with

```java
       @Override
        public EdgeIteratorState set(BooleanEncodedValue property, boolean value) {
            property.setBool(reverse, getFlags(), value);
            if (baseNode == adjNode) {
                property.setBool(!reverse, getFlags(), value);
            }
            baseGraph.writeFlags(edgePointer, getFlags());
            return this;
        }

        @Override
        public EdgeIteratorState setReverse(BooleanEncodedValue property, boolean value) {
            property.setBool(!reverse, getFlags(), value);
            if (baseNode == adjNode) {
                property.setBool(reverse, getFlags(), value);
            }
            baseGraph.writeFlags(edgePointer, getFlags());
            return this;
        }
```     

?

**Comment by easbar** [comment]
... or we even reject setReverse entirely for loops?

**Comment by easbar** [comment]
We have a few options:

1) Silently set the opposite direction's value in set/setReverse as proposed above. I do not like this so much because if we do
```java
loop.set(property, true);
loop.setReverse(property, false);
```
we get `loop.get(property) == false` (even though we just set it to `true`.

2) Reject `setReverse` for loops altogether. This way we are forced to check if an edge is a loop before calling `setReverse` which seems reasonable to me. But what about `getReverse`? Again we could reject it for loops, but then we even have to check for loops when calling the getter. This seems cumbersome to me. We could forward `loop.getReverse()` to `loop.get()`, or set the reverse property also when we call `loop.set()`. The usage would be like this:

```java
loop.set(property, true); // ok
loop.setReverse(property, true); // error
loop.get(property); // -> true
loop.getReverse(property); // -> true
```

Any opinions about this? I have a small preference for the second option.

**Comment by karussell** [comment]
Yes, second options sounds reasonable. And you would then remove the special handling for loops on import? 

> loop.setReverse(property, true); // error

Probably only an error if the property has support for one value per direction?

**Comment by easbar** [comment]
> Probably only an error if the property has support for one value per direction?

This is another good question. Currently we have this in `set(property, boolean fwd, boolean bwd)`:
```
            if (!property.isStoreTwoDirections())
                throw new IllegalArgumentException("EncodedValue " + property.getName() + " supports only one direction");
```
But why is it not also an error to call `setReverse` when `!property.isStoreTwoDirections`. Isn't it like when we only have one direction there is no 'reverse' property so we can only use get/set (and not getReverse/setReverse)?

> Yes, second options sounds reasonable. And you would then remove the special handling for loops on import?

Ok and yes hopefully we can remove the special handling then. However, we probably need another special handling then. In case there are actual loops in OSM with different speeds for the two directions or different geometries etc. (does this exist?) we actually have to add two loops (one for each direction). And right now I am wondering if this can cause other problems, like turning from one loop to the other etc.

**Comment by easbar** [comment]
>  we actually have to add two loops (one for each direction).

Probably we need to do this for *all* loops, because otherwise we won't be able to e.g. block one direction dynamically.

**Comment by karussell** [comment]
> But why is it not also an error to call setReverse when !property.isStoreTwoDirections.

We can try this and see if test succeed. Maybe the problem was that our code is in general "direction-dependent" and calls setReverse explicitly even if not stored separately.

> maybe we can just as well split all loop ways into two edges like we currently do and stop using loops in BaseGraph altogether?!

Ok. If it makes so many problems and we already avoid this in the import, why not :) ?

**Comment by easbar** [comment]
> Ok. If it makes so many problems and we already avoid this in the import, why not :) ?

I put in a lot of effort to make e.g. edge-based CH work with base-graph loops (shortcut loops are always necessary). And for all I know it is working, so we solved a lot of problems related to such loops already and maybe we can handle loops. But with my current knowledge I think directed loops are ill-defined so there is still some work to get this right. At the same time right now I do not know about a single 'real' use-case for loop edges and this somewhat suggests that we maybe should not spend time dealing with loops. Loop ways in OSM can be split into two edges. Loop edges used to represent stations in PT are only used because LocationIndex can only find edges, but these can be replaced by small loops as well. Can you think of other potential use-cases for loops?

**Comment by easbar** [comment]
Just to be clear what the real issue is:

1) We allow setting different speeds, access etc. for the two directions of each edge (including loops)
2) When we traverse edges using `EdgeExplorer` we process one edge ID at a time. Each edge ID represents two directions (two edge keys). The routing algorithms typically only consider one of these directions (only outgoing edges for the forward search, only incoming edges for the backward search), i.e. they call `get(speedEnc)` or `getReverse(speedEnc)` but not both. 
3) For loops we need to consider both directions. This is why we already adjusted the `DefaultEdgeFilter` so it accepts loops if only one of the two directions is accessible: https://github.com/graphhopper/graphhopper/blob/e9a2f9b6c12dd0f95be197bcb5ccc62091f36077/core/src/main/java/com/graphhopper/routing/util/DefaultEdgeFilter.java#L58-L68 However, given only an edgeID it is unclear which direction should be considered to calculate e.g. the weight. 
4) For the weight calculation we blindly call get/getReverse for the forward/backward search and since for loops we always get the same direction in some cases we get the wrong one. Actually there is not even a 'correct' direction and for loops we would have to consider both directions or take the minimum of the forward and backward weight.

I think it is not worth it to mess around with loops any longer and we should just remove them from our graph model altogether.
